### DISCUSSION TITLE
Alternative to fringes as section visibility indicators

### DISCUSSION THREAD

**Discussion by bkaestner** [magit/magit#4909] [Ideas]
At the moment, `magit-section-visibility-indicator` understands either bitmaps for fringes or a single ellipsis string. I would like to propse a third choice, namely `(STRING . STRING)` as an indicator, where the `car` is used as `string-prefix` for closed sections and the `cdr` is used as `string-prefix` for open sections (see "Envisioned Result" below for an example)

### Motivation
The fringes look OK on a regular display, but on a high DPI display they are almost unnoticeable:

![grafik](https://user-images.githubusercontent.com/6301773/227792958-0e2b8356-741b-4482-890b-67715c1ffdfa.png)

Note the giant size mismatch between the fringe and the font itself. This is with `(face-attribute 'default :height)` equal to `99`.

### Envisioned Result
With `(setq magit-section-visibility-indicator (cons "▶" "▼"))`:

![grafik](https://user-images.githubusercontent.com/6301773/227792917-18210c69-9a34-446f-8a23-d8a403ff35c2.png)

With the `Head` section being closed:

![grafik](https://user-images.githubusercontent.com/6301773/227793212-a632c390-32de-4eb4-9682-e2ea36148c09.png)

### Proof-of-concept code
In `magit-section-maybe-update-visibility-indicator`, add a new `cond` section:
```lisp
(defun magit-section-maybe-update-visibility-indicator (section)
  (when (and magit-section-visibility-indicator
             (magit-section-content-p section))
    (let* ((beg (oref section start))
           (eoh (save-excursion
                  (goto-char beg)
                  (line-end-position))))
      (cond
       ((symbolp (car-safe magit-section-visibility-indicator))
        ...)
       ;; Patch starts here
       ((stringp (cdr-safe magit-section-visibility-indicator))
        (let ((ov (magit--overlay-at beg 'magit-vis-indicator 'beg)))
          (unless ov
            (setq ov (make-overlay beg (1+ beg)))
            (overlay-put ov 'evaporate t)
            (overlay-put ov 'magit-vis-indicator 'beg))
          (overlay-put ov 'before-string
                       (concat
                        (if (oref section hidden)
                            (car magit-section-visibility-indicator)
                          (cdr magit-section-visibility-indicator))
                        " "))))
       ;; Patch ends here
       ((stringp (car-safe magit-section-visibility-indicator))
        ...)))))
```
I'm not entirely sure whether the `cdr` breaks something else though, but I would be happy to provide this as a PR. :)

**Comment by tarsius** [comment]
Another solution would be to make the fringe wider (globally or just in magit buffers) and to use larger bitmaps. That's the solution I had in mind when I made it possible to select the bitmaps to be used.

Even better, the bitmaps could be generated programmatically based on the pixel height of the default face and the pixel width of the left fringe. (That's what I really wanted, but I was hoping someone else would come along and implement it. Maybe there is already an implementation of that to be found somewhere, maybe even in Emacs itself.)

**Reply by bkaestner** [reply]
Thanks for the comment Jonas!

> Another solution would be to make the fringe wider (globally or just in magit buffers) and to use larger bitmaps.

That would definitely help with the fringes on high DPI displays.

> Even better, the bitmaps could be generated programmatically based on the pixel height of the default face and the pixel width of the left fringe.

That sounds somewhat difficult and error-prone, to be honest, but I get why you would prefer that compared to an overlay at the beginning of the line. That being said, would you be interested in a proper PR for a visibility indicator at the beginning of the line in addition to the current one at the end of the line? Or should I put that into a new 3rd party package?

Thanks again for your time!

**Reply by tarsius** [reply]
The indicators can now be shown in the margin. Customize `magit-section-visibility-indicators` to enable that. I don't intend to implement support for showing the indicator in the buffer.
